node/src/components/block_proposer/deploy_sets.rs

```diff
@@ -11,12 +11,21 @@ use crate::{
     types::{DeployHash, DeployHeader, Timestamp},
     Chainspec,
 };
+use casper_execution_engine::shared::gas::Gas;
+
+// Key for pending collection: willingness to pay (header.gas_price), header.payment_amount(),
+// deploy_hash
+type PendingKey = (u64, Gas, DeployHash);
 
 /// Stores the internal state of the BlockProposer.
 #[derive(Clone, DataSize, Debug, Deserialize, Serialize)]
 pub struct BlockProposerDeploySets {
     /// The collection of deploys pending for inclusion in a block.
-    pub(super) pending: HashMap<DeployHash, DeployType>,
+    /// The key of this map is used as a method of keeping the map sorted first by `gas_price`
+    /// (willingness to pay), then by `payment_amount` and finally include `deploy_hash` to
+    /// facilitate map operations.
+    /// https://github.com/CasperLabs/ceps/blob/Gas_spot_market/text/0022-gas-spot-market.md#ordering
+    pub(super) pending: BTreeMap<PendingKey, DeployType>,
```

Comment:
A `BinaryHeap` might be slightly faster, and would avoid the need to duplicate the data in `PendingKey`. It also has a `retain` method that would simplify `prune_pending_deploys`.

Reply:
I guess it's a question of optimizing for deletion or for pruning. I could see how this could be done with an `Ord` impl and `BinaryHeap`, but one wrinkle in that plan is that `std::collections::BinaryHeap::retain` is still nightly-only. There's a `BinaryHeapPlus` crate that I was looking at, but I opted to go for the `BTreeMap` approach because it more resembled the existing `HashMap` interface that was here before.